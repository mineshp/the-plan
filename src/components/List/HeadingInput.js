import React from 'react';
import PropTypes from 'prop-types';
import { Button, Divider, Form, Grid, Input } from 'semantic-ui-react';

const HeadingInput = ({ headings, handleAddColumn, handleRemoveColumn, handleHeaderInputChange }) => {
    const headerRows = [];
    headings.map((heading) => {
        const headingID = heading.id; // eslint-disable-line no-underscore-dangle
        return (
            headerRows.push(
                <Grid.Row key={headingID}>
                    <Grid.Column>
                        <Form.Field>
                            <Input
                                placeholder="Header Name..."
                                name={heading.name}
                                id={headingID}
                                key={headingID}
                                defaultValue={heading.name}
                                onChange={handleHeaderInputChange}
                            />
                        </Form.Field>
                    </Grid.Column>
                    <Grid.Column>
                        <Button
                            icon="trash"
                            content="Remove"
                            labelPosition="left"
                            id={headingID}
                            color="pink"
                            type="button"
                            onClick={handleRemoveColumn}
                            floated="left"
                        />
                    </Grid.Column>
                </Grid.Row>
            )
        );
    });

    return (
        <div>
            <Grid columns={2} divided>
                <Divider horizontal>ASSIGN LIST HEADINGS</Divider>
                {headerRows}
                <Divider />
                <Button fluid type="button" color="blue" onClick={handleAddColumn}>Add Column</Button>
            </Grid>
        </div>
    );
};

HeadingInput.propTypes = {
    headings: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string,
            name: PropTypes.string
        }).isRequired
    ).isRequired,
    handleAddColumn: PropTypes.func.isRequired,
    handleRemoveColumn: PropTypes.func.isRequired,
    handleHeaderInputChange: PropTypes.func.isRequired
};

export default HeadingInput;
