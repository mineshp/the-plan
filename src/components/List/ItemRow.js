import React from 'react';
import PropTypes from 'prop-types';
import { Button, Table } from 'semantic-ui-react';


const ItemRow = ({ data, listID }) => (
    <Table.Row>
        <Table.Cell>{data.name}</Table.Cell>
        <Table.Cell>{data.createdDate}</Table.Cell>
        <Table.Cell>{data.updatedDate}</Table.Cell>
        <Table.Cell>{data.notes}</Table.Cell>
        <Table.Cell>{data.owner}</Table.Cell>
        <Table.Cell><Button as='a' color='teal' href={`/list/${listID}/item/${data.id}/delete`}>Delete</Button></Table.Cell>
    </Table.Row>
);

ItemRow.propTypes = {
  data: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    createdDate: PropTypes.string.isRequired,
    updatedDate: PropTypes.string.isRequired
    }).isRequired
}

export default ItemRow;