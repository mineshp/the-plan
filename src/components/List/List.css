.ui.input.stdInput {
    width: 90% !important;
}

.ui.formSpacing {
    margin-top: 40px;
}