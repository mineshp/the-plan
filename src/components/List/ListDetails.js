import React from 'react';
import PropTypes from 'prop-types';
import Table from 'semantic-ui-react/dist/commonjs/collections/Table';
import { formatDate } from '../../helpers/validators/common';

const ListDetails = ({ list }) => {
    const owner = list.owner ? list.owner : 'unknown';
    return (
        <Table>
            <Table.Body>
                <Table.Row>
                    <Table.Cell className="details-header">Owner:</Table.Cell>
                    <Table.Cell>{owner}</Table.Cell>
                    <Table.Cell className="details-header">Created Date:</Table.Cell>
                    <Table.Cell>{formatDate(list.createdDate)}</Table.Cell>
                    <Table.Cell className="details-header">Last Updated:</Table.Cell>
                    <Table.Cell>{formatDate(list.updatedDate)}</Table.Cell>
                </Table.Row>
            </Table.Body>
        </Table>
    );
};

ListDetails.propTypes = {
    list: PropTypes.shape({}).isRequired
};

export default ListDetails;
