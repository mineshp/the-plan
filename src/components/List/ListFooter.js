import React from 'react';
import PropTypes from 'prop-types';
import { Table, Button, Icon } from 'semantic-ui-react';

const ListFooter = ({ numColumns, handleAddItem, handleSubmit }) => (
    <Table.Footer fullWidth>
        <Table.Row>
            <Table.HeaderCell colSpan={numColumns}>
                <Button floated="left" color="blue" icon labelPosition="left" size="small" onClick={handleAddItem}>
                    <Icon name="plus" /> Add Item
                </Button>
                <Button
                    floated="right"
                    icon
                    labelPosition="left"
                    color="green"
                    size="small"
                    onClick={handleSubmit}
                >
                    <Icon name="save" /> Save
                </Button>
            </Table.HeaderCell>
        </Table.Row>
    </Table.Footer>
);

ListFooter.propTypes = {
    numColumns: PropTypes.number.isRequired,
    handleAddItem: PropTypes.func.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default ListFooter;
