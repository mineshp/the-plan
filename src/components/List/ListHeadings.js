import React from 'react';
import PropTypes from 'prop-types';
import { Button, Dropdown, Table } from 'semantic-ui-react';

const ListItems = ({ downloadPDF, headings, listId }) => {
    const columnHeadings = [];
    headings.map((heading) => (
        columnHeadings.push(<Table.HeaderCell key={heading.id}>{heading.name}</Table.HeaderCell>)
    ));

    columnHeadings.push(
        <Table.HeaderCell key={'actions'} textAlign="right" colSpan="2">
            <Button
                as="a"
                icon="edit"
                color="green"
                href={`/list/update/${listId}`}
                title="Amend List Setup"
            />
            <Dropdown text="Export PDF" icon="file pdf outline" floating labeled button className="icon pdf">
                <Dropdown.Menu>
                    <Dropdown.Item id={listId} onClick={downloadPDF}>Save</Dropdown.Item>
                    <Dropdown.Item disabled>Email</Dropdown.Item>
                </Dropdown.Menu>
            </Dropdown>
        </Table.HeaderCell>
    );

    return (
        <Table.Header>
            <Table.Row key={123}>
                {columnHeadings}
            </Table.Row>
        </Table.Header>
    );
};

ListItems.propTypes = {
    headings: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            position: PropTypes.number
        })
    ).isRequired,
    downloadPDF: PropTypes.func.isRequired,
    listId: PropTypes.string.isRequired
};

ListItems.defaultProps = {
    headings: []
};

export default ListItems;
