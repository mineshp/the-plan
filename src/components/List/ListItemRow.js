import React from 'react';
import PropTypes from 'prop-types';
import { Button, Input, Table } from 'semantic-ui-react';

const calculateColumnWidths = (numOfCols) => {
    const MAX_WIDTH = 12;
    const DEL_BTN_WIDTH = 1;
    const COMPLETED_BTN_WIDTH = 1;

    return Math.ceil(MAX_WIDTH / (numOfCols + DEL_BTN_WIDTH + COMPLETED_BTN_WIDTH));
};

const ListItemRow = ({ handleChange, handleDelete, handleCompleted, itemRow }) => {
    const row = [];
    const rowProps = {};
    const markAsCompletedClass = itemRow.completed ? 'completed-item' : '';
    itemRow.columns.map((column) => {
        const itemInputKey = `${itemRow.rowId}-${column.columnName}`;
        rowProps.positive = itemRow.completed;
        return row.push(
            <Table.Cell key={itemInputKey} width={calculateColumnWidths(itemRow.columns.length)} {...rowProps}>
                <Input
                    fluid
                    id={itemRow.rowId}
                    name={column.columnName}
                    value={column.columnValue}
                    onChange={handleChange}
                    className={markAsCompletedClass}
                    disabled={rowProps.positive}
                />
            </Table.Cell>
        );
    });

    const completedItemRowBtnKey = `${itemRow.rowId}-completedRow`;
    const markItemAsCompleteBtn = (
        <Button
            icon="thumbs up"
            color="yellow"
            id={itemRow.rowId}
            onClick={handleCompleted}
        />
    );

    const markItemAsUnCompleteBtn = (
        <Button
            icon="check"
            color="green"
            id={itemRow.rowId}
            onClick={handleCompleted}
        />
    );

    const completedBtn = itemRow.completed ? markItemAsUnCompleteBtn : markItemAsCompleteBtn;

    row.push(
        <Table.Cell key={completedItemRowBtnKey} width="1" textAlign="right" {...rowProps}>
            { completedBtn }
        </Table.Cell>
    );

    const deleteRowBtnKey = `${itemRow.rowId}-deleteRow`;
    row.push(
        <Table.Cell key={deleteRowBtnKey} width="1" textAlign="right" {...rowProps}>
            <Button
                icon="trash"
                color="pink"
                id={itemRow.rowId}
                onClick={handleDelete}
            />
        </Table.Cell>
    );

    return (
        <Table.Row>
            {row}
        </Table.Row>
    );
};

ListItemRow.propTypes = {
    itemRow: PropTypes.shape({}).isRequired,
    handleChange: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleCompleted: PropTypes.func.isRequired
};

export default ListItemRow;
