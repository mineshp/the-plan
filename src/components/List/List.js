import React from 'react';
import PropTypes from 'prop-types';
import { Accordion, Container, Header, Icon, Table } from 'semantic-ui-react';
import ListHeadings from './ListHeadings';
import ListItems from './ListItems';
import ListFooter from './ListFooter';
import ListDetails from './ListDetails';
import ListUploadFile from './ListUploadFile';

const List = ({
    downloadPDF, handleAddItem, handleChange, handleDelete,
    handleSubmit, handleCompleted, items, list, activeIndex,
    handleListAccordionClick
}) => {
    const numColumns = list.headings.length + 2;
    const listId = list._id; // eslint-disable-line no-underscore-dangle
    return (
        <div className="List main content-body">
            <Container>
                <Header as="h1">{list.listName}</Header>
                <ListDetails list={list} />
                <Table striped>
                    <ListHeadings downloadPDF={downloadPDF} listId={listId} headings={list.headings} />
                    <ListItems
                        items={items}
                        handleChange={handleChange}
                        handleDelete={handleDelete}
                        handleCompleted={handleCompleted}
                    />
                    <ListFooter
                        numColumns={numColumns}
                        handleAddItem={handleAddItem}
                        handleSubmit={handleSubmit}
                    />
                </Table>

                <Accordion fluid styled>
                    <Accordion.Title active={activeIndex === 0} id="0" onClick={handleListAccordionClick}>
                        <Icon name="dropdown" />
                            View Images for List
                    </Accordion.Title>
                    <Accordion.Content active={activeIndex === 0}>
                        AAA
                    </Accordion.Content>

                    <Accordion.Title active={activeIndex === 1} id="1" onClick={handleListAccordionClick}>
                        <Icon name="dropdown" />
                            Upload Image
                    </Accordion.Title>
                    <Accordion.Content active={activeIndex === 1}>
                        <ListUploadFile />
                    </Accordion.Content>
                </Accordion>
            </Container>
        </div>
    );
};

List.propTypes = {
    list: PropTypes.shape({
        _id: PropTypes.string.isRequired,
        headings: PropTypes.arrayOf(PropTypes.shape({})).isRequired
    }).isRequired,
    items: PropTypes.arrayOf(
        PropTypes.shape({})
    ).isRequired,
    downloadPDF: PropTypes.func.isRequired,
    handleAddItem: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    handleCompleted: PropTypes.func.isRequired,
    handleListAccordionClick: PropTypes.func.isRequired,
    activeIndex: PropTypes.number.isRequired
};

List.defaultProps = {
    list: {
        projects: [],
        headings: [],
        items: [],
        rows: []
    }
};

export default List;
