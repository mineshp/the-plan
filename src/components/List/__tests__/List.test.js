import React from 'react';
import { shallow } from 'enzyme';
import List from '../List';
import ListHeadingComponent from '../ListHeadings';
import ListItemsComponent from '../ListItems';
import ListFooterComponent from '../ListFooter';
import { mockSingleList, mockItemsData } from '../../../helpers/test/testData/listData';

const mockHandleAddItem = jest.fn();
const mockHandleChange = jest.fn();
const mockHandleSubmit = jest.fn();
const mockHandleDelete = jest.fn();
const mockItems = mockItemsData();
const mockListData = mockSingleList();
const mockDownloadPDF = jest.fn();

describe('List', () => {
    it('renders ListHeadings component', () => {
        const wrapper = shallow(<List
            list={mockListData}
            handleAddItem={mockHandleAddItem}
            handleChange={mockHandleChange}
            handleSubmit={mockHandleSubmit}
            handleDelete={mockHandleDelete}
            downloadPDF={mockDownloadPDF}
            items={mockItems}
        />);
        expect(wrapper.find(ListHeadingComponent)).toHaveLength(1);
    });

    it('renders ListItemsComponent component', () => {
        const wrapper = shallow(<List
            list={mockListData}
            handleAddItem={mockHandleAddItem}
            handleChange={mockHandleChange}
            handleSubmit={mockHandleSubmit}
            handleDelete={mockHandleDelete}
            downloadPDF={mockDownloadPDF}
            items={mockItems}
        />);
        expect(wrapper.find(ListItemsComponent)).toHaveLength(1);
    });

    it('renders ListFooterComponent component', () => {
        const wrapper = shallow(<List
            list={mockListData}
            handleAddItem={mockHandleAddItem}
            handleChange={mockHandleChange}
            handleSubmit={mockHandleSubmit}
            handleDelete={mockHandleDelete}
            downloadPDF={mockDownloadPDF}
            items={mockItems}
        />);
        expect(wrapper.find(ListFooterComponent)).toHaveLength(1);
    });
});
