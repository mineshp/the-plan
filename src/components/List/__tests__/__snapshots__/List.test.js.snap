// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders list correctly 1`] = `
<div
  className="List main"
>
  <div
    className="ui container"
  >
    <h1
      className="ui header"
    >
      Avengers
    </h1>
    <table
      className="ui striped table"
    >
      <thead
        className=""
      >
        <tr
          className=""
        >
          <th
            className=""
          >
            a
          </th>
          <th
            className=""
          >
            b
          </th>
          <th
            className=""
          >
            c
          </th>
          <th
            className="right aligned"
          >
            <div
              aria-busy={undefined}
              aria-disabled={undefined}
              aria-expanded={false}
              aria-multiselectable={undefined}
              className="ui button floating labeled dropdown icon pdf"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              >
                Export PDF
              </div>
              <i
                aria-hidden="true"
                className="file pdf outline icon"
                onClick={[Function]}
              />
              <div
                className="menu transition"
              >
                <div
                  aria-checked={undefined}
                  aria-disabled={undefined}
                  aria-selected={undefined}
                  className="item"
                  id="123"
                  onClick={[Function]}
                  role="option"
                >
                  Save
                </div>
                <div
                  aria-checked={undefined}
                  aria-disabled={true}
                  aria-selected={undefined}
                  className="disabled item"
                  onClick={[Function]}
                  role="option"
                >
                  Email
                </div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        className=""
      >
        <tr
          className=""
        >
          <td
            className="six wide"
          >
            <div
              className="ui fluid input"
            >
              <input
                disabled={undefined}
                id="001-002"
                name="Name"
                onChange={[Function]}
                tabIndex={undefined}
                type="text"
                value="Captain America"
              />
            </div>
          </td>
          <td
            className="six wide"
          >
            <div
              className="ui fluid input"
            >
              <input
                disabled={undefined}
                id="001-002"
                name="Desc"
                onChange={[Function]}
                tabIndex={undefined}
                type="text"
                value="The first avenger"
              />
            </div>
          </td>
          <td
            className="right aligned six wide"
          >
            <button
              className="ui pink icon button"
              disabled={undefined}
              id="001-002"
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              <i
                aria-hidden="true"
                className="trash icon"
              />
            </button>
          </td>
        </tr>
      </tbody>
      <tfoot
        className="full-width"
      >
        <tr
          className=""
        >
          <th
            className=""
            colSpan={4}
          >
            <button
              className="ui blue small icon left floated left labeled button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              <i
                aria-hidden="true"
                className="plus icon"
              />
               Add Item
            </button>
            <button
              className="ui green small icon right floated left labeled button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              <i
                aria-hidden="true"
                className="save icon"
              />
               Save
            </button>
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
